### PR TITLE
Similarity db generator

### PR DESCRIPTION
**Pull request by pvinchon**
A single executable to to run image similarity. This will fetch or generate the required dependencies.

The image cropping has been removed for two main reasons:
- it required human interaction
- the result from a cropped and uncropped run are similar
- changing from dress to [shoes, shirts, handbags, ...] would require retraining the cropping

Will be looking into injecting image paths next. `ROOT/resources/dresses-db/master` is quite limitting

### CODE REVIEW

File: src/torch/9-similarity-db/similarity-db.lua [hunk 1]
```diff
@@ -38,12 +38,6 @@ function similarityDb()
       similarities[test] = similarity or -1
     end
 
-    -- compare itself with its mirror
-    local flipped = tiefvision_commons.dataPath('encoded-images-flipped', reference)
-    local flippedEncoding = torch.load(flipped):double()
-    local similarity = similarity_lib.similarity(referenceEncoding, flippedEncoding)
-    similarities[reference] = similarity or -1
-
```
Comment (paucarre) [inline src/torch/9-similarity-db/similarity-db.lua]: Why is this removed?

File: src/torch/8-similarity-db-cnn/generate-similarity-db.lua [hunk 2]
```diff
@@ -30,5 +33,4 @@ function createDb(sourceFolder, destinationFolder)
   end
 end
 
-createDb(tiefvision_commons.resourcePath('dresses-db/bboxes/1'), tiefvision_commons.dataPath('encoded-images'))
-createDb(tiefvision_commons.resourcePath('dresses-db/bboxes-flipped/1'), tiefvision_commons.dataPath('encoded-images-flipped'))
```
Comment (paucarre) [inline src/torch/8-similarity-db-cnn/generate-similarity-db.lua]: Why is this removed?

File: src/torch/9-similarity-db/similarity-db.lua (hunk 1 above)
Comment (Plippe) [inline src/torch/9-similarity-db/similarity-db.lua]: As explained in the PR description, the similarities generated with the whole images are comparable to those generated with the cropped images. By removing the cropping process, we gain the possibility of running the similarity with any type of images without any manual intervention. 

Flipped images not being used by the unsupervised learning, except for comparing an image to itself, I saw little points in keeping them.

**Comment by paucarre** [comment]
I don't agree we should remove cropping the image location part, that's really important for similarity. For example in terms of removing noise due to legs and shoes and filtering out dress lengths.
We can do it later on if we add a spatial transformer.
So for me removing the bounding box detection is nok.

**Comment by paucarre** [comment]
A few points more:
1- Human interaction will be necessary for supervised similarity, which we will have to most likely do before production.
2- Bounding box for any class can be automated adding a spatial transformer for the similarity network which will imply training it with a dataset made by humans
3- If we want to add just unsupervised learning without bounding box nor anything else, then we could have gone with the features extracted from tensorflow model with just a few modifications.

**Comment by Plippe** [comment]
The idea of removing the cropping and replacing it with custom paths allow people to run similarity on what ever type of image they wish. If cropping is deemed important, images can be cropped, and the script can be pointed to those cropped images. If cropping isn't a requirement, this step can be skipped. 

The cropping is currently a task that is specific to category and requires human intervention. Those two points make it unpractical for production use.

Lastly, here are two similarities, one with cropping, and one without. I am unable to tell which is best.

screen shot 2016-12-01 at 15 09 54

screen shot 2016-12-01 at 15 10 13

**Comment by Plippe** [comment]
Offline discussion points to the need to implement an `input` and `output` flag. This flag should be used if the default app behavior isn't wished. The flags' default values should be the current values.

**Review by paucarre — APPROVED** [review]
